File: ports/cpp/test/whitebox/WhiteboxTest.cpp
#include <ParserRuleContext.h>
#include <WhiteboxLexer.h>
#include <WhiteboxParser.h>
#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include <antlr4-c3/CodeCompletionCore.hpp>
#include <cstdlib>
#include <utility/AntlrPipeline.hpp>
#include <utility/Collections.hpp>
#include <utility/Testing.hpp>

namespace c3::test {

struct WhiteboxGrammar {
  using Lexer = WhiteboxLexer;
  using Parser = WhiteboxParser;
};

/// (optional tokens)
TEST(WhiteboxGrammarTests, CaretAtTransitionToRuleWithNonExhaustiveFollowSet) {
  AntlrPipeline<WhiteboxGrammar> pipeline("LOREM ");
  auto* ctx = pipeline.parser.test1();
  EXPECT_EQ(pipeline.listener.GetErrorCount(), 1);

  c3::CodeCompletionCore completion(&pipeline.parser);
  auto candidates = completion.collectCandidates(1, {ctx});

  EXPECT_THAT(
      Keys(candidates.tokens),
      UnorderedElementsAre(
          WhiteboxLexer::IPSUM,
          WhiteboxLexer::DOLOR,
          WhiteboxLexer::SIT,
          WhiteboxLexer::AMET,
          WhiteboxLexer::CONSECTETUR
      )
  );
}

/// (epsilon-only transition to rule end)
TEST(WhiteboxGrammarTests, CaretAtTransitionToRuleWithEmptyFollowSet) {
  AntlrPipeline<WhiteboxGrammar> pipeline("LOREM ");
  auto* ctx = pipeline.parser.test2();
  EXPECT_EQ(pipeline.listener.GetErrorCount(), 1);

  c3::CodeCompletionCore completion(&pipeline.parser);
  auto candidates = completion.collectCandidates(1, {ctx});

  EXPECT_THAT(
      Keys(candidates.tokens),
      UnorderedElementsAre(
          WhiteboxLexer::IPSUM,
          WhiteboxLexer::DOLOR,
          WhiteboxLexer::SIT,
          WhiteboxLexer::AMET,
          WhiteboxLexer::CONSECTETUR
      )
  );
}

TEST(WhiteboxGrammarTests, CaretAtOneOfMultiplePossibleStates) {
  for (const auto index : {4, 5, 6, 7}) {
    AntlrPipeline<WhiteboxGrammar> pipeline("LOREM IPSUM ");

    auto* ctx = [&]() -> antlr4::ParserRuleContext* {
      switch (index) {
        case 4:
          return pipeline.parser.test4();
        case 5:  // NOLINT: magic
          return pipeline.parser.test5();
        case 6:  // NOLINT: magic
          return pipeline.parser.test6();
        case 7:  // NOLINT: magic
          return pipeline.parser.test7();
        default:
          std::abort();
      }
    }();

    c3::CodeCompletionCore completion(&pipeline.parser);
    auto candidates = completion.collectCandidates(2, {ctx});

    EXPECT_THAT(Keys(candidates.tokens), UnorderedElementsAre(WhiteboxLexer::DOLOR));
    EXPECT_THAT(candidates.tokens[WhiteboxLexer::DOLOR], ElementsAre());
  }
}

TEST(WhiteboxGrammarTests, CaretAtOneOfMultiplePossibleStatesWithCommonFollowList) {
  AntlrPipeline<WhiteboxGrammar> pipeline("LOREM IPSUM ");

  auto* ctx = pipeline.parser.test8();

  c3::CodeCompletionCore completion(&pipeline.parser);
  auto candidates = completion.collectCandidates(2, {ctx});

  EXPECT_THAT(Keys(candidates.tokens), UnorderedElementsAre(WhiteboxLexer::DOLOR));
  EXPECT_THAT(candidates.tokens[WhiteboxLexer::DOLOR], ElementsAre(WhiteboxLexer::SIT));
}

}  // namespace c3::test

File: ports/java/src/test/java/com/vmware/antlr4c3/WhiteboxTest.java
package com.vmware.antlr4c3;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.util.List;
import java.util.Set;

import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.CharStreams;
import org.antlr.v4.runtime.CommonTokenStream;
import org.antlr.v4.runtime.ParserRuleContext;
import org.junit.Test;

public class WhiteboxTest {
    @Test
    public void caretAtTransitionToRuleWithNonExhaustiveFollowSetTest() {
        CharStream inputStream = CharStreams.fromString("LOREM ");

        WhiteboxLexer lexer = new WhiteboxLexer(inputStream);
        CommonTokenStream tokenStream = new CommonTokenStream(lexer);

        WhiteboxParser parser = new WhiteboxParser(tokenStream);
        CountingErrorListener errorListener = new CountingErrorListener();
        parser.removeErrorListeners();
        parser.addErrorListener(errorListener);

        ParserRuleContext ctx = parser.test1();

        assertEquals(1, errorListener.errorCount);

        CodeCompletionCore core = new CodeCompletionCore(parser, null, null);
        CodeCompletionCore.CandidatesCollection candidates = core.collectCandidates(1, ctx);

        assertEquals(5, candidates.tokens.size());
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.IPSUM));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.DOLOR));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.SIT));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.AMET));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.CONSECTETUR));
    }

    @Test
    public void caretAtTransitionToRuleWithEmptyFollowSetTest() {
        CharStream inputStream = CharStreams.fromString("LOREM ");
        WhiteboxLexer lexer = new WhiteboxLexer(inputStream);
        CommonTokenStream tokenStream = new CommonTokenStream(lexer);

        WhiteboxParser parser = new WhiteboxParser(tokenStream);
        CountingErrorListener errorListener = new CountingErrorListener();
        parser.removeErrorListeners();
        parser.addErrorListener(errorListener);

        ParserRuleContext ctx = parser.test2();
        assertEquals(1, errorListener.errorCount);

        CodeCompletionCore core = new CodeCompletionCore(parser, null, null);
        CodeCompletionCore.CandidatesCollection candidates = core.collectCandidates(1, ctx);

        assertEquals(5, candidates.tokens.size());
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.IPSUM));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.DOLOR));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.SIT));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.AMET));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.CONSECTETUR));
    }

    @Test
    public void caretAtOptionalTokenTest() {
        CharStream inputStream = CharStreams.fromString("LOREM ");
        WhiteboxLexer lexer = new WhiteboxLexer(inputStream);
        CommonTokenStream tokenStream = new CommonTokenStream(lexer);

        WhiteboxParser parser = new WhiteboxParser(tokenStream);
        CountingErrorListener errorListener = new CountingErrorListener();
        parser.removeErrorListeners();
        parser.addErrorListener(errorListener);

        ParserRuleContext ctx = parser.test3();
        assertEquals(1, errorListener.errorCount);

        CodeCompletionCore core = new CodeCompletionCore(parser, null, null);
        CodeCompletionCore.CandidatesCollection candidates = core.collectCandidates(1, ctx);

        assertEquals(4, candidates.tokens.size());
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.IPSUM));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.DOLOR));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.SIT));
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.AMET));
    }

    @Test
    public void caretAtOneOfMultiplePossibleStatesTest() {
        for (int index : new int[] { 4, 5, 6, 7 }) {
            CharStream inputStream = CharStreams.fromString("LOREM IPSUM ");
            WhiteboxLexer lexer = new WhiteboxLexer(inputStream);
            CommonTokenStream tokenStream = new CommonTokenStream(lexer);

            WhiteboxParser parser = new WhiteboxParser(tokenStream);
            ParserRuleContext ctx;
            switch (index) {
                case 4:
                    ctx = parser.test4();
                    break;
                case 5:
                    ctx = parser.test5();
                    break;
                case 6:
                    ctx = parser.test6();
                    break;
                case 7:
                    ctx = parser.test7();
                    break;
                default:
                    throw new IllegalStateException();
            }

            CodeCompletionCore core = new CodeCompletionCore(parser, null, null);
            CodeCompletionCore.CandidatesCollection candidates = core.collectCandidates(2, ctx);

            assertEquals(Set.of(WhiteboxLexer.DOLOR), candidates.tokens.keySet());
            assertTrue(candidates.tokens.get(WhiteboxLexer.DOLOR).isEmpty());
        }
    }

    @Test
    public void caretAtOneOfMultiplePossibleStatesWithCommonFollowListTest() {
        CharStream inputStream = CharStreams.fromString("LOREM IPSUM ");
        WhiteboxLexer lexer = new WhiteboxLexer(inputStream);
        CommonTokenStream tokenStream = new CommonTokenStream(lexer);

        WhiteboxParser parser = new WhiteboxParser(tokenStream);
        ParserRuleContext ctx = parser.test8();

        CodeCompletionCore core = new CodeCompletionCore(parser, null, null);
        CodeCompletionCore.CandidatesCollection candidates = core.collectCandidates(2, ctx);

        assertEquals(1, candidates.tokens.size());
        assertTrue(candidates.tokens.containsKey(WhiteboxLexer.DOLOR));
        assertEquals(List.of(WhiteboxLexer.SIT), candidates.tokens.get(WhiteboxLexer.DOLOR));
    }
}
